Packaging tgz outputs a zip file, not a gzipped tarball

Commit 913010720ddf26008bb769170ef3a86677e5e31c forces compress mode of zip for all artifacts, even those that are not zip files. The original packaging from the options needs to be passed as the mode to the grunt-contrib-compress options

Package created with the following grunt configs produces a `.tgz` file which was not a valid `.tgz` format, unable to unpack with `maven` as it complains not a valid format.

```javascript
 maven: {
        options: {
            groupId: 'com.sample.hello',
            artifactId: 'hello-world',
            packaging: 'tgz',
            classifier: 'distrib',
            injectDestFolder: 'false'
        },
        package: {
            goal: 'package',
            expand: true,
            cwd: 'dist/',
            src: ['**/*'],
            dest: '/',
            dist: 'target'
        }
}
```

Maven unable to unpack:

```sh
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-dependency-plugin:2.10:unpack (unpack) on project sample: 
Error unpacking file: 
/Users/sample/.m2/repository/com/sample/hello/hello-world/1.0.0-SNAPSHOT/hellow-world-1.0.0-SNAPSHOT-distrib.tgz  to: /Users/something/

[ERROR] org.codehaus.plexus.archiver.ArchiverException: 
Error while expanding /Users/sample/.m2/repository/com/sample/hello/hello-world/1.0.0-SNAPSHOT/hellow-world-1.0.0-SNAPSHOT-distrib.tgz: Not in GZIP format
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,9 +1,10 @@
  language: node_js
  sudo: false
  node_js:
-   - "4.0"
+   - "node"
+   - "5"
+   - "4"
    - "0.10"
-   - "0.8"
  before_install:
   - npm update -g npm
  before_script:

--- a/package.json
+++ b/package.json
@@ -28,7 +28,7 @@
   ],
   "main": "Gruntfile.js",
   "engines": {
-    "node": ">= 0.8.0"
+    "node": ">= 0.10.0"
   },
   "scripts": {
     "test": "grunt test"

--- a/package.json
+++ b/package.json
@@ -1,23 +1,29 @@
 {
   "name": "grunt-maven-tasks",
   "description": "Grunt plugin to deploy a file to a maven repository",
-  "version": "1.3.7",
-  "homepage": "https://github.com/smh/grunt-maven-tasks",
+  "version": "1.3.8",
+  "homepage": "https://github.com/jpmeyer/grunt-maven-tasks",
   "author": {
     "name": "Stein Martin Hustad",
     "email": "stein@hustad.com"
   },
+  "contributors": [
+    {
+      "name": "John Meyer",
+      "email": "johnpmeyer@gmail.com"
+    }
+  ],
   "repository": {
     "type": "git",
-    "url": "https://github.com/smh/grunt-maven-tasks.git"
+    "url": "https://github.com/jpmeyer/grunt-maven-tasks.git"
   },
   "bugs": {
-    "url": "https://github.com/smh/grunt-maven-tasks/issues"
+    "url": "https://github.com/jpmeyer/grunt-maven-tasks/issues"
   },
   "licenses": [
     {
       "type": "MIT",
-      "url": "https://github.com/smh/grunt-maven-tasks/blob/master/LICENSE-MIT"
+      "url": "https://github.com/jpmeyer/grunt-maven-tasks/blob/master/LICENSE-MIT"
     }
   ],
   "main": "Gruntfile.js",
@@ -28,8 +34,8 @@
     "test": "grunt test"
   },
   "dependencies": {
-    "semver": "~2.3.1",
-    "grunt-contrib-compress": "~0.11.0"
+    "semver": ">5.1.0",
+    "grunt-contrib-compress": "1.3.x"
   },
   "devDependencies": {
     "mocha": "~1.21.4",
@@ -43,7 +49,7 @@
     "jszip": "~2.5.0"
   },
   "peerDependencies": {
-    "grunt": "~0.4.5"
+    "grunt": ">=0.4.5"
   },
   "keywords": [
     "gruntplugin",

--- a/package.json
+++ b/package.json
@@ -34,7 +34,7 @@
     "test": "grunt test"
   },
   "dependencies": {
-    "semver": ">5.1.0",
+    "semver": ">=5.1.0",
     "grunt-contrib-compress": "1.3.x"
   },
   "devDependencies": {
@@ -46,7 +46,7 @@
     "grunt": "~0.4.5",
     "grunt-contrib-jshint": "~0.10.0",
     "grunt-simple-mocha": "git+https://github.com/yaymukund/grunt-simple-mocha.git#7d1cb9bfd8580fc5369bd82e288f683bb7ca0d46",
-    "jszip": "~2.5.0"
+    "jszip": "~3.0.0"
   },
   "peerDependencies": {
     "grunt": ">=0.4.5"

--- a/tasks/maven_tasks.js
+++ b/tasks/maven_tasks.js
@@ -164,7 +164,7 @@ module.exports = function(grunt) {
     
     options.packaging = getExtension(options.packaging, options.classifier, options.type);
 
-    grunt.config.set('maven.package.options', { archive: options.file, mode: 'zip', extension: options.packaging });
+    grunt.config.set('maven.package.options', { archive: options.file, mode: ['tar', 'tgz', 'gzip', 'gz'].indexOf(options.packaging) > 0 ? options.packaging : 'zip', extension: options.packaging });
     grunt.config.set('maven.package.files', task.files);
     grunt.config.set('maven.deploy-file.options', options);
     grunt.config.set('maven.install-file.options', options);

--- a/tasks/maven_tasks.js
+++ b/tasks/maven_tasks.js
@@ -161,10 +161,8 @@ module.exports = function(grunt) {
   }
 
   function configureMaven(options, task) {
-    
     options.packaging = getExtension(options.packaging, options.classifier, options.type);
-
-    grunt.config.set('maven.package.options', { archive: options.file, mode: ['tar', 'tgz', 'gzip', 'gz'].indexOf(options.packaging) > 0 ? options.packaging : 'zip', extension: options.packaging });
+    grunt.config.set('maven.package.options', { archive: options.file, extension: options.packaging });
     grunt.config.set('maven.package.files', task.files);
     grunt.config.set('maven.deploy-file.options', options);
     grunt.config.set('maven.install-file.options', options);
@@ -172,9 +170,13 @@ module.exports = function(grunt) {
 
   grunt.registerTask('mvn:package', function() {
     var done = this.async();
-      
+    var options = grunt.config('maven.package.options');
+    if(!options.mode) {
+      var packaging = options.packaging || options.extension;
+      options.mode = (packaging === 'tar' || packaging === 'tgz' || packaging === 'gzip' || packaging === 'gz' || packaging === 'tar.gz') ? packaging : 'zip';
+    }
     var compress = require('grunt-contrib-compress/tasks/lib/compress')(grunt);
-    compress.options = grunt.config('maven.package.options');
+    compress.options = options;
     compress.tar(grunt.config('maven.package.files'), function(){
       renameForKnownPackageTypeArtifacts(compress.options.archive, compress.options.extension);
       done();

--- a/test/package-test.js
+++ b/test/package-test.js
@@ -17,12 +17,11 @@ var initConfig = {
 maven: {
   options: {
       groupId: 'test.project', 
-      type: 'war'
     },
-    'package': {
+    package: {
       options: {
         goal: 'package',
-	src: [ '**', '!node_modules/**' ],
+        src: [ '**', '!node_modules/**' ]
       }
     }
   }
@@ -33,11 +32,11 @@ maven: {
   testPackage(target, 'package.json', { name: 'test-project', version: '1.0.0' });
   testPackage(target, 'package.json', { name: 'test-project', version: '1.0.0-SNAPSHOT', packaging: 'war' });
   testPackage(target, 'package.json', { name: 'test-project', version: '1.0.0', packaging: 'zip' });
-  testPackage(target, 'package.json', { name: 'test-project', version: '1.0.0-SNAPSHOT', packaging: 'tgz' });
+  testPackage(target, 'package.json', { name: 'test-project', version: '1.0.0-SNAPSHOT', packaging: 'tgz', mode: 'tgz'});
 });
 
 function testPackage(target, versionFile, pkg) {
-  describe(target + ' - ' + pkg.version + ':' + pkg.classifier + ' -', function() {
+  describe(target + ' - ' + pkg.version + ':' + pkg.classifier + (pkg.packaging ? ':' + pkg.packaging : '') + ' -', function() {
     before(function(done) {
       async.series([
         function(cb) { setupGruntProject(versionFile, pkg, initConfig, cb); },
@@ -49,7 +48,7 @@ function testPackage(target, versionFile, pkg) {
     });
 
     it('should create a package file', function(done) {
-      verifyPackageFile(pkg.name, pkg.version, pkg.classifier, pkg.packaging, done);
+      verifyPackageFile(pkg.name, pkg.version, pkg.classifier, pkg.packaging, pkg.mode, done);
     });
   });
 }
@@ -64,15 +63,35 @@ function exec(command, fn) {
   });
 }
 
-function verifyPackageFile(project, version, classifier, packaging, cb) {
+function verifyPackageFile(project, version, classifier, packaging, mode, cb) {
   var filename = project + '-' + version + (classifier? '-' + classifier : '') + '.' + getExtension(packaging, classifier, initConfig.maven.options.type);
 
   fs.readFile(path.join(projectDir, filename),
       function(err, data) {
-	if (err) { throw err; }
-	var zip = new jszip(data);
-	cb();
-      });
+	if (err) { cb(err); }
+    switch(mode) {
+      case 'tgz':
+        if(!data) {
+          cb('File data is null');
+        } else if (data.length < 3) {
+          cb('tgz package content is too small: ' + data.length);
+        } else if (data[0] !== 0x1f) {
+          cb('Expect first byte[0x1f], got ' + data[0]);
+        } else if (data[1] !== 0x8b) {
+          cb('Expect first byte[0x8b], got ' + data[1]);
+        } else if (data[2] !== 0x08) {
+          cb('Expect first byte[0x08], got ' + data[2]);
+        } else {
+          cb();
+        }
+        break;
+      default:
+        jszip.loadAsync(data).then(function() {
+          cb();
+        }, cb);
+        break;
+    }
+  });
 }
 
 // copy from maven_task

--- a/test/zip-test.js
+++ b/test/zip-test.js
@@ -69,15 +69,15 @@ function verifyZipFile(pkg, cb) {
   fs.readFile(path.join(projectDir, deploy.file),
       function(err, data) {
         if (err) {
-          throw err;
+         cb(err);
         }
 
         jszip.loadAsync(data).then(function(zip) {
           var access = (zip.files[pkg.name + '-' + pkg.version + '/' + relScriptFile].unixPermissions & 511).toString(8);
           access.should.equal('755'); // When (the correct) grunt-contrib-compress 0.7 or higher is used.
           cb();
-        }, function() {
-          throw 'Invalid zip file';
+        }, function(e) {
+          cb(e);
         });
       });
 }

--- a/test/zip-test.js
+++ b/test/zip-test.js
@@ -68,14 +68,17 @@ function verifyZipFile(pkg, cb) {
 
   fs.readFile(path.join(projectDir, deploy.file),
       function(err, data) {
-	if (err) {
-    throw err;
-  }
-	var zip = new jszip(data);
-	var access=(zip.files[pkg.name +'-'+pkg.version + '/' + relScriptFile].unixPermissions & 511).toString(8);
-        access.should.equal('755'); // When (the correct) grunt-contrib-compress 0.7 or higher is used.
-	//access.should.equal('0'); // When grunt-contrib-compress 0.4 is used.
-	cb();
+        if (err) {
+          throw err;
+        }
+
+        jszip.loadAsync(data).then(function(zip) {
+          var access = (zip.files[pkg.name + '-' + pkg.version + '/' + relScriptFile].unixPermissions & 511).toString(8);
+          access.should.equal('755'); // When (the correct) grunt-contrib-compress 0.7 or higher is used.
+          cb();
+        }, function() {
+          throw 'Invalid zip file';
+        });
       });
 }
 

